compiler/rustc_ast/src/mut_visit.rs

```diff
@@ -286,8 +286,11 @@ pub trait MutVisitor: Sized {
         walk_where_clause(self, where_clause);
     }
 
-    fn visit_where_predicate(&mut self, where_predicate: &mut WherePredicate) {
-        walk_where_predicate(self, where_predicate);
+    fn filter_map_where_predicate(
+        &mut self,
+        where_predicate: WherePredicate,
+    ) -> Option<WherePredicate> {
```

Comment:
Is it because `flat_map` reserves the space that we can easily support macros which produce multiple `where` predicates in place?

Except for that reason, I don't think there are any differences between `filter_map` and `flat_map` in terms of time or memory efficiency. `FlatMapInPlace::flat_map_in_place` works in the same way for `Option<WherePredicate>` and `SmallVec<[WherePredicate; 1]>` that contains at most one item; and they both have the same memory size (`size_of::<WherePredicate>() + size_of::<usize>()`).